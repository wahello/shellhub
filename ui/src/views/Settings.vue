<template>
  <v-container>
    <v-card
      :elevation="0"
    >
      <v-app-bar
        flat
        color="transparent"
      >
        <v-tabs
          centered
        >
          <v-tab
            v-for="item in visibleItems"
            :key="item.title"
            :to="item.path"
            :data-test="item.title+'-tab'"
          >
            {{ item.title }}
          </v-tab>
        </v-tabs>
      </v-app-bar>

      <v-divider />

      <v-container
        class="pa-0"
        fluid
      >
        <router-view />
      </v-container>
    </v-card>
  </v-container>
</template>

<script>

export default {
  name: 'Settings',

  data() {
    return {
      drawer: true,
      clipped: false,
    };
  },

  computed: {
    visibleItems() {
      return this.items.filter((item) => !item.hidden);
    },

    currentInANamespace() {
      return localStorage.getItem('tenant') !== '';
    },

    hasNamespace() {
      return this.$store.getters['namespaces/getNumberNamespaces'] !== 0;
    },

    items() {
      return [
        {
          title: 'Profile',
          path: '/settings',
        },
        {
          title: 'Namespace',
          path: '/settings/namespace-manager',
          hidden: !this.currentInANamespace,
        },
        {
          title: 'Private Keys',
          path: '/settings/private-keys',
        },
        {
          title: 'Tags',
          path: '/settings/tags',
        },
        {
          title: 'Billing',
          path: '/settings/billing',
          hidden: !(this.$env.isCloud && this.hasNamespace),
        },
      ];
    },
  },
};

</script>

<style>
.v-list-active {
  border-left: 4px solid var(--v-primary-base);
}

.text-shadow {
  text-shadow: #000 0 0 6px;
  color: transparent;
}
</style>
