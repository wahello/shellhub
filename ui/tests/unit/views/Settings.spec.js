import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import Settings from '@/views/Settings';

describe('Settings', () => {
  const localVue = createLocalVue();
  localVue.use(VueRouter);
  localVue.use(Vuex);

  let wrapper;

  let numberNamespaces = 1;

  const items = [
    {
      title: 'Profile',
      path: '/settings',
    },
    {
      title: 'Namespace',
      path: '/settings/namespace-manager',
    },
    {
      title: 'Private Keys',
      path: '/settings/private-keys',
    },
    {
      title: 'Tags',
      path: '/settings/tags',
    },
    {
      title: 'Billing',
      path: '/settings/billing',
    },
  ];

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      numberNamespaces,
    },
    getters: {
      'namespaces/getNumberNamespaces': (state) => state.numberNamespaces,
    },
    actions: {
    },
  });

  ///////
  // In this case, the billing tab can only be rendered in the cloud.
  // Checks if this tab has been rendered when user has namespace.
  ///////

  describe('Cloud is true', () => {
    beforeEach(() => {
      wrapper = shallowMount(Settings, {
        localVue,
        store,
        stubs: ['fragment'],
        mocks: {
          $env: {
            isCloud: true,
          },
        },
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.drawer).toEqual(true);
      expect(wrapper.vm.clipped).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      Object.keys(items).forEach((item) => {
        expect(wrapper.find(`[data-test="${items[item].title}-tab"]`).text()).toEqual(items[item].title);
      });
    });
  });

  ///////
  // In this case, the billing tab can only be rendered in the cloud.
  // Checks if this tab is not rendered when user has no namespace.
  ///////

  describe('Cloud is false', () => {
    numberNamespaces = 0;

    beforeEach(() => {
      wrapper = shallowMount(Settings, {
        localVue,
        stubs: ['fragment'],
        mocks: {
          $env: {
            isCloud: false,
          },
        },
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.drawer).toEqual(true);
      expect(wrapper.vm.clipped).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      Object.keys(items.slice(0, -1)).forEach((item) => {
        expect(wrapper.find(`[data-test="${items[item].title}-tab"]`).text()).toEqual(items[item].title);
      });
    });
  });

  ///////
  // In this case, the billing tab can only be rendered in the cloud.
  // Checks if this tab is not rendered.
  ///////

  describe('Cloud is false', () => {
    beforeEach(() => {
      wrapper = shallowMount(Settings, {
        localVue,
        stubs: ['fragment'],
        mocks: {
          $env: {
            isCloud: false,
          },
        },
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.drawer).toEqual(true);
      expect(wrapper.vm.clipped).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      Object.keys(items.slice(0, -1)).forEach((item) => {
        expect(wrapper.find(`[data-test="${items[item].title}-tab"]`).text()).toEqual(items[item].title);
      });
    });
  });
});
