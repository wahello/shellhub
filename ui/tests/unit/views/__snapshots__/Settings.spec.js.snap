// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Cloud is false Renders the component 1`] = `
<v-container-stub tag="div">
  <v-card-stub loaderheight="4" tag="div" elevation="0">
    <v-app-bar-stub color="transparent" tag="header" extensionheight="48" flat="true" src="" value="true">
      <v-tabs-stub activeclass="" centered="true" nexticon="$next" previcon="$prev" slidersize="2">
        <v-tab-stub to="/settings" ripple="true" data-test="Profile-tab">
          Profile
        </v-tab-stub>
        <v-tab-stub to="/settings/namespace-manager" ripple="true" data-test="Namespace-tab">
          Namespace
        </v-tab-stub>
        <v-tab-stub to="/settings/private-keys" ripple="true" data-test="Private Keys-tab">
          Private Keys
        </v-tab-stub>
        <v-tab-stub to="/settings/tags" ripple="true" data-test="Tags-tab">
          Tags
        </v-tab-stub>
      </v-tabs-stub>
    </v-app-bar-stub>
    <v-divider-stub></v-divider-stub>
    <v-container-stub tag="div" fluid="true" class="pa-0">
      <router-view-stub name="default"></router-view-stub>
    </v-container-stub>
  </v-card-stub>
</v-container-stub>
`;

exports[`Settings Cloud is false Renders the component 2`] = `
<v-container-stub tag="div">
  <v-card-stub loaderheight="4" tag="div" elevation="0">
    <v-app-bar-stub color="transparent" tag="header" extensionheight="48" flat="true" src="" value="true">
      <v-tabs-stub activeclass="" centered="true" nexticon="$next" previcon="$prev" slidersize="2">
        <v-tab-stub to="/settings" ripple="true" data-test="Profile-tab">
          Profile
        </v-tab-stub>
        <v-tab-stub to="/settings/namespace-manager" ripple="true" data-test="Namespace-tab">
          Namespace
        </v-tab-stub>
        <v-tab-stub to="/settings/private-keys" ripple="true" data-test="Private Keys-tab">
          Private Keys
        </v-tab-stub>
        <v-tab-stub to="/settings/tags" ripple="true" data-test="Tags-tab">
          Tags
        </v-tab-stub>
      </v-tabs-stub>
    </v-app-bar-stub>
    <v-divider-stub></v-divider-stub>
    <v-container-stub tag="div" fluid="true" class="pa-0">
      <router-view-stub name="default"></router-view-stub>
    </v-container-stub>
  </v-card-stub>
</v-container-stub>
`;

exports[`Settings Cloud is true Renders the component 1`] = `
<v-container-stub tag="div">
  <v-card-stub loaderheight="4" tag="div" elevation="0">
    <v-app-bar-stub color="transparent" tag="header" extensionheight="48" flat="true" src="" value="true">
      <v-tabs-stub activeclass="" centered="true" nexticon="$next" previcon="$prev" slidersize="2">
        <v-tab-stub to="/settings" ripple="true" data-test="Profile-tab">
          Profile
        </v-tab-stub>
        <v-tab-stub to="/settings/namespace-manager" ripple="true" data-test="Namespace-tab">
          Namespace
        </v-tab-stub>
        <v-tab-stub to="/settings/private-keys" ripple="true" data-test="Private Keys-tab">
          Private Keys
        </v-tab-stub>
        <v-tab-stub to="/settings/tags" ripple="true" data-test="Tags-tab">
          Tags
        </v-tab-stub>
        <v-tab-stub to="/settings/billing" ripple="true" data-test="Billing-tab">
          Billing
        </v-tab-stub>
      </v-tabs-stub>
    </v-app-bar-stub>
    <v-divider-stub></v-divider-stub>
    <v-container-stub tag="div" fluid="true" class="pa-0">
      <router-view-stub name="default"></router-view-stub>
    </v-container-stub>
  </v-card-stub>
</v-container-stub>
`;
